### PR TITLE
Move `WCOrderFetcher` to FluxC. Fetch outdated and missing orders together.

### LINKED ISSUE #2012: Fetch outdated *and* missing orders when summaries has been fetched
**Context**

I'm working on improving how `WCAndroid` handles new order notification. After discussion in #2007 , I've decided to, for now, fetch the list of orders using a diff mechanism.

**Problems to resolve in the scope of this issue**

1. Currently, we check for missing orders in [`OrderListItemDataSource#getItemsAndFetchIfNecessary`](https://github.com/woocommerce/woocommerce-android/blob/f70c1e79ebca097119b7c3b6bef65648239a058d/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListItemDataSource.kt#L57-L60). This is problematic as **it requires `OrderListFragment` to exist**, which is not a case when an app is not running but notification is received (in `Service`). This is described in https://github.com/woocommerce/woocommerce-android/issues/4053

2. If there is a case when the app has some missing **and** outdated orders, the **app will request an update twice**. For the first time, [`OrderListItemDataSource#getItemsAndFetchIfNecessary`](https://github.com/woocommerce/woocommerce-android/blob/f70c1e79ebca097119b7c3b6bef65648239a058d/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListItemDataSource.kt#L57-L60) for missing orders and [`WCStoreOrder#handleFetchOrderListCompleted`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/5f2fb2f652ec711d17c5b7392c54491b1259f370/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderStore.kt#L578) for outdated ones. 

**Solution**

The solution I came by is to move `WCOrderFetcher` from `WCAndroid` to `FluxC` and request missing orders at the same time as we request outdated ones - in `WCOrderStore`.

2  Crash to Sentry sequence(1)
(I don't specify **_something_** as it's not important for the scope of this issue. Now it's `MainPresenter` or `OrderListFragment`, later it'll be `NotificationHandler`)

**Other ideas I've tried**
1. Just requesting an orders update (without moving `WCOrderFetcher` to `FluxC`). It won't work because of duplicated events. We could remove `WCOrderFetcher#fetch` from [`OrderListItemDataSource#getItemsAndFetchIfNecessary`](https://github.com/woocommerce/woocommerce-android/blob/f70c1e79ebca097119b7c3b6bef65648239a058d/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListItemDataSource.kt#L57-L60) but I don't want to do this - it's I think safer to have an additional check (see diagram).

2. Fetching a single order (with e.g. `FETCH_ORDERS_BY_IDS` action), we won't have a `summary` model created, so the result won't appear on the list.

3. Fetching a single order with creating a `summary` model is a big effort. Described in #2007 .

### PR DESCRIPTION
**Pull request by wzieba**
Resolves: #2012 

This PR:

1. Moves updated `WCOrderFetcher` from `WCAndroid` to `FluxC`. 
2. Makes `WCOrderStore` request for missing orders in addition to the request for outdated ones.

Please check #2012 for the motivation behind those changes.

### CODE REVIEW

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderFetcher.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * This Singleton is a light-weight class that provides methods for ensuring that orders matching a provided list
+ * of [RemoteId]'s exist in the local DB. If they do not exist, or if they have been modified since last saved to
+ * the DB, FluxC will pull down fresh versions from the API and save them to the DB. Once all orders have been
+ * successfully fetched, any [androidx.paging.PagedList] views observing this specific type of data will update
+ * automatically.
+ */
+@Singleton
+class WCOrderFetcher @Inject constructor(private val dispatcher: Dispatcher) {
```
Comment (wzieba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderFetcher.kt:26]: This implementation is the same as the actual [`WCAndroid` implementation](https://github.com/woocommerce/woocommerce-android/blob/0821552ebc66e4f352c00e8046382979731e72d2/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderFetcher.kt).

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderStore.kt [hunk 2]
```diff
@@ -593,14 +595,34 @@ class WCOrderStore @Inject constructor(dispatcher: Dispatcher, private val wcOrd
...
-        mDispatcher.dispatch(WCOrderActionBuilder.newFetchOrdersByIdsAction(payload))
+    private fun fetchOutdatedOrMissingOrders(site: SiteModel, fetchedSummaries: List<WCOrderSummaryModel>) {
+        val fetchedSummariesIds = fetchedSummaries.map { RemoteId(it.remoteOrderId) }
+        val localOrdersForFetchedSummaries = OrderSqlUtils.getOrdersForSiteByRemoteIds(site, fetchedSummariesIds)
+
+        val idsToFetch = outdatedOrdersIds(fetchedSummaries, localOrdersForFetchedSummaries)
+                .plus(missingOrdersIds(fetchedSummariesIds, localOrdersForFetchedSummaries))
+
+        wcOrderFetcher.fetchOrders(site = site, remoteItemIds = idsToFetch)
+    }
```
Comment (wzieba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderStore.kt:606]: I plan to add tests to this.

**Comment by wzieba** [comment]
You can preview how it works with `WCAndroid` branch `issue/4053-fetch-order-for-incomming-notification`.

**Comment by wzieba** [comment]
@oguzkocer @AmandaRiu as main contributors of `ListStore` and `OrderFetcher`, can I ask you for checking the motivation behind this change in #2012 and initial review of this draft?

I didn't yet finalize this PR (there's no test or `WCAndroid` PR) but before doing that, I'd like to get your feedback if this change seems ok to you and the principles of the diff mechanism. Thank you 🙏 .

**Comment by oguzkocer** [comment]
@wzieba It looks like, besides moving a few things around, this is just adding a way to detect which orders are not yet fetched and combines that with which orders are outdated to fetch them together. That all seems very reasonable to me and I thought it already worked this way [as we do something very similar in PostStore.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/store/PostStore.java#L870-L930) (albeit that one looks like a huge wall of text since it's in Java and there is a few more tricky things for posts)

I'll leave the actual review to @AmandaRiu but I think this looks great - thank you for making it happen!

**Comment by oguzkocer** [comment]
@wzieba I am taking myself from the review requests. It looks like I didn't use the review feature to comment on this, so I am still assigned. Please feel free to add me again if you'd like me to take another look at it.

**Comment by wzieba** [comment]
Thank you @oguzkocer for your input!

> (...) this is just adding a way to detect which orders are not yet fetched and combines that with which orders are outdated to fetch them together.

Exactly this ☝️ . Thank you for sharing `PostStore` snippet, it looks indeed similar.

> I'll leave the actual review to @AmandaRiu but I think this looks great - thank you for making it happen!

Thanks!

@AmandaRiu just a heads up that I plan to add tests and prepare a `WCAndroid` PR early in the next week.

**Comment by wzieba** [comment]
hi, @AmandaRiu 👋 on the second thought I think that it'll be beneficial for the Woo team to review this PR too so I'll request a review from the whole team. But if you have time, it'd be great to have your review too. Thanks!

**Comment by nbradbury** [comment]
@wzieba As with @oguzkocer, this all seems reasonable to me. The logic is simply moved from WCAndroid to FluxC, where it should be. It gets a 👍 from me, but I would be more comfortable if @AmandaRiu took a  look because she's **much** more familiar with this code than I am.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderStore.kt [hunk 3]
```diff
@@ -593,14 +595,34 @@ class WCOrderStore @Inject constructor(dispatcher: Dispatcher, private val wcOrd
...
+        }.map(WCOrderModel::remoteOrderId).map(::RemoteId)
+    }
+
+    private fun missingOrdersIds(
+        fetchedSummariesIds: List<RemoteId>,
+        localOrdersForSiteByRemoteIds: List<WCOrderModel>
+    ): List<RemoteId> {
+        return fetchedSummariesIds.minus(
+                localOrdersForSiteByRemoteIds.map(WCOrderModel::remoteOrderId).map(::RemoteId)
+        )
```
Comment (wzieba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderStore.kt:625]: Those are crucial methods. Request to reviewers: please take a second look at those, I've unit tested it but would love to have a second look.

**Comment by wzieba** [comment]
@AmandaRiu do you think you'll have some capacity to review this PR? Or is it okay to merge this with just approves from members of @wordpress-mobile/woo-android team?

Edit: we'll go ahead and merge it but if there are any concerns about this change, please free to do a post-mortem review and I'll happily check that. Thank you!

**Review by kidinov — APPROVED** [review]
I was reviewing mostly in `WCOrderStore`. Those changes look logical to me =)

**Pull request merged**